### PR TITLE
make torch.testing error messages more expressive

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56373 add `torch.testing` to docs
* #56365 Split `test_testing.py::TestAsserts` for multiple devices
* #55890 enable support for custom error messages in `torch.testing`
* #55788 add support for equal_nan in torch.testing.assert_close
* #55786 Add support for scalars and numpy in torch.testing
* #55385 Add support for checking tensor containers in `torch.testing`
* **#55145 make torch.testing error messages more expressive**

Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111

A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

Differential Revision: [D27877157](https://our.internmc.facebook.com/intern/diff/D27877157)

### CODE REVIEW

**Comment by pmeier** [comment]
From @mruberry in https://github.com/pytorch/pytorch/pull/54784#issuecomment-816385781

> Brief elaboration on this topic:
>
> It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs. Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.

**Comment by pmeier** [comment]
> It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs.

Not sure what you mean there. Isn't the index doing exactly that? In the example above `a[2, 3]` and `b[2, 3]` is the combination that results in the maximum absolute and relative difference.

> Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.

I've removed just that, albeit private, in this PR: `_trace_mismatches`. Do you want something like that as public interface? If so, how would it look like? `trace_(equal|close)_mismatches(a, b, **kwargs)`? What would it return? A named tuple with all the information?

**Comment by mruberry** [comment]
> > It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs.
> 
> Not sure what you mean there. Isn't the index doing exactly that? In the example above `a[2, 3]` and `b[2, 3]` is the combination that results in the maximum absolute and relative difference.

Yes but it'd be nice to do that progammatically so the test could acquire the inputs and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".

> > Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.
> 
> I've removed just that, albeit private, in this PR: `_trace_mismatches`. Do you want something like that as public interface? If so, how would it look like? `trace_(equal|close)_mismatches(a, b, **kwargs)`? What would it return? A named tuple with all the information?

I'm not sure yet. Let's focus on getting the core right and then we can consider additions like this.

File: test/test_testing.py [hunk 1]
```diff
@@ -798,42 +798,80 @@ def test_mismatching_values(self, device):
 
     @onlyCPU
     def test_mismatching_values_msg_abs_mismatches(self, device):
-        a = torch.empty((3, 4), dtype=torch.float32, device=device).fill_(5)
+        a = torch.empty((3, 3), dtype=torch.float32, device=device).fill_(5)
```
Comment (mruberry) [inline test/test_testing.py:801]: Prefer zeros to empty, since empty could be anything (including the correct value)

File: torch/testing/_asserts.py [hunk 2]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
+    """Traces mismatches.
 
     Args:
         a (torch.Tensor): First tensor.
         b (torch.Tensor): Second tensor.
         mismatches (torch.Tensor): Boolean mask of the same shape as :attr:`a` and :attr:`b` that indicates the
             location of mismatches.
+
+    Returns:
+        (str): Multiline list comprising the absolute and relative number of mismatches, and the maximum absolute and
```
Comment (mruberry) [inline torch/testing/_asserts.py:167]: A string that contains the greatest absolute difference between elements of a and b, the greatest relative difference between elements of a and b, and the indices of these differences.

File: torch/testing/_asserts.py [hunk 3]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
+            relative difference with the corresponding index.
     """
-    total = mismatches.numel()
-    abs = torch.sum(mismatches).item()
-    rel = abs / total
+    total_elements = mismatches.numel()
+    abs_mismatches = torch.sum(mismatches).item()
+    rel_mismatches = abs_mismatches / total_elements
```
Comment (mruberry) [inline torch/testing/_asserts.py:172]: Because "relative" is so overloaded this should be the percentage of mismatched elements

File: torch/testing/_asserts.py [hunk 4]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
+
+    Returns:
+        (str): Multiline list comprising the absolute and relative number of mismatches, and the maximum absolute and
+            relative difference with the corresponding index.
     """
-    total = mismatches.numel()
-    abs = torch.sum(mismatches).item()
-    rel = abs / total
+    total_elements = mismatches.numel()
+    abs_mismatches = torch.sum(mismatches).item()
```
Comment (mruberry) [inline torch/testing/_asserts.py:171]: Because "absolute" is so overloaded this would be the total number of mismatched elements

File: torch/testing/_asserts.py [hunk 5]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
+    rel_mismatches = abs_mismatches / total_elements
 
     dtype = torch.float64 if a.dtype.is_floating_point else torch.int64
     a_flat = a.flatten().to(dtype)
```
Comment (mruberry) [inline torch/testing/_asserts.py:175]: Why flatten?

File: torch/testing/_asserts.py [hunk 6]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
-        b=b_flat[idx_flat].item(),
+    abs_diff = torch.abs(a_flat - b_flat)
+    max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
+
+    rel_diff = abs_diff / torch.abs(b_flat)
+    max_rel_diff, max_rel_diff_flat_idx = torch.max(rel_diff, 0)
+
+    return (
+        f"Mismatched elements: {abs_mismatches} / {total_elements} ({rel_mismatches:.1%})\n"
+        f"Max. abs. diff.: {max_abs_diff.item()} at {_unravel_index(max_abs_diff_flat_idx.item(), mismatches.shape)}\n"
```
Comment (mruberry) [inline torch/testing/_asserts.py:186]: "Greatest absolute difference"

File: torch/testing/_asserts.py [hunk 7]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
+        f"Max. rel. diff.: {max_rel_diff.item()} at {_unravel_index(max_rel_diff_flat_idx.item(), mismatches.shape)}"
```
Comment (mruberry) [inline torch/testing/_asserts.py:187]: "Greatest relative difference:"

File: torch/testing/_asserts.py [hunk 8]
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
-        abs=abs,
-        rel=rel,
-        idx=_unravel_index(idx_flat, a.shape),
-        diff=abs_diff_flat[idx_flat].item(),
-        a=a_flat[idx_flat].item(),
-        b=b_flat[idx_flat].item(),
+    abs_diff = torch.abs(a_flat - b_flat)
+    max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
+
+    rel_diff = abs_diff / torch.abs(b_flat)
```
Comment (mruberry) [inline torch/testing/_asserts.py:181]: This is an interesting way of computing the greatest relative difference, because it ignores the magnitude of elements of a. 

What if a is large but b is small? Won't that report a huge relative error but if a is small and b is large the relative error would be small?

Can we consider being symmetric? I understand that NumPy's actual comparison operations are asymmetric. I wonder if that's changing in the array API?

File: torch/testing/_asserts.py [hunk 9]
```diff
@@ -243,11 +227,8 @@ def _check_values_close(
     if not torch.any(mismatches):
         return None
 
-    trace = _trace_mismatches(a, b, mismatches)
     return AssertionError(
-        f"With rtol={rtol} and atol={atol}, "
-        f"found {trace.abs} different element(s) out of {trace.total} ({trace.rel:.1%}). "
-        f"The greatest difference of {trace.diff} ({trace.a} vs. {trace.b}) occurred at index {trace.idx}"
+        f"Tensors are not close according to rtol={rtol} and ator={atol}:\n\n{_trace_mismatches(a, b, mismatches)}"
```
Comment (mruberry) [inline torch/testing/_asserts.py:231]: "ator" -> "atol"

Comment (mruberry) [reply torch/testing/_asserts.py:231]: "Tensors are not close:"

or

"Tensors are not close!" 

(see above)

File: torch/testing/_asserts.py (hunk 6 above)
Comment (mruberry) [inline torch/testing/_asserts.py:186]: It might be interesting to return these statistics as data. Then the close assertion could create its own string that inserted the greatest absolute difference and greatest relative difference that would be allowed.

File: test/test_testing.py (hunk 1 above)
Comment (pmeier) [reply test/test_testing.py:801]: 1. Shouldn't this make no difference given that we are invoking `fill_()` after the instantiation?
2. Why do you want to change the size?

File: torch/testing/_asserts.py (hunk 3 above)
Comment (pmeier) [reply torch/testing/_asserts.py:172]: But percentage would also not be right unless we multiply by 100 here, right?

File: torch/testing/_asserts.py (hunk 5 above)
Comment (pmeier) [reply torch/testing/_asserts.py:175]: To be able to easily find the maximum. [`torch.max`](https://pytorch.org/docs/stable/generated/torch.max.html#torch-max) operates on one dimension. If we would pass in an input with more dimensions we would get a dimension-wise maximum here.

File: torch/testing/_asserts.py (hunk 8 above)
Comment (pmeier) [reply torch/testing/_asserts.py:181]: Note, that [`torch.isclose`](https://pytorch.org/docs/stable/generated/torch.isclose.html?highlight=isclose#torch.isclose) and thus in turn `assert_tensors_close` is also asymmetric. This way `rel_diff` is the minimal value for `rtol` that would make the check pass.

> Can we consider being symmetric?

Sure, how would you go about it? Compute the relative difference both ways and pick the smaller one?

> I understand that NumPy's actual comparison operations are asymmetric.

Not sure what you mean by "actual comparison". If you have a look at [`numpy.testing.assert_allclose`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy-testing-assert-allclose) they are using the same logic to find mismatches. They named the parameters `actual` and `desired` to make this more clear. 

I'll check if they do anything different for their error message.

**Comment by pmeier** [comment]
> Yes but it'd be nice to do that progammatically so the test could acquire the indices and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".

We could add a `msg` parameter to our public interface. If it is used, every error message will simply replaced by it. 

To give the user more options, we could also accept a callable that will be called with the mismatch statistics and returns the message of the error to be raised.

File: torch/testing/_asserts.py (hunk 6 above)
Comment (pmeier) [reply torch/testing/_asserts.py:186]: So something like https://github.com/pytorch/pytorch/pull/55145#issuecomment-817297273?

**Comment by mruberry** [comment]
> > Yes but it'd be nice to do that progammatically so the test could acquire the indices and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".
> 
> We could add a `msg` parameter to our public interface. If it is used, every error message will simply replaced by it.
> 
> To give the user more options, we could also accept a callable that will be called with the mismatch statistics and returns the message of the error to be raised.

Maybe. Let's talk about it more for a future change.

File: torch/testing/_asserts.py (hunk 8 above)
Comment (mruberry) [reply torch/testing/_asserts.py:181]: You're right. I guess if we keep it this way then the atol and rtol are just these values, because the check is consistent with close.

What about "actual" and "expected" for names? It'd be nice to clarify to users which tensor should go where.

File: torch/testing/_asserts.py (hunk 9 above)
Comment (mruberry) [reply torch/testing/_asserts.py:231]: You're correct that if we follow the current isclose() logic we don't need to take this recommendation, because the atol and rtol needed are just the greatest absolute difference and greatest relative difference. Maybe instead we can say:

"Greatest absolute difference: {x} (up to {y} allowed)"

or:

"
Greatest absolute difference: {x}
Allowed absolute difference: {y}
"

File: torch/testing/_asserts.py (hunk 3 above)
Comment (pmeier) [reply torch/testing/_asserts.py:172]: I went for `mismatch_ratio` instead.

File: torch/testing/_asserts.py (hunk 8 above)
Comment (pmeier) [reply torch/testing/_asserts.py:181]: Can do that. Do you want this naming scheme only for `assert_close` or everywhere for consistency? `assert_equal` is symmetric.

File: test/test_testing.py (hunk 1 above)
Comment (mruberry) [reply test/test_testing.py:864]: Good point. I missed the fill_, sorry.

In that case uses torch.full().

I'm not sure I understand what you mean about changing the size?

File: torch/testing/_asserts.py (hunk 8 above)
Comment (mruberry) [reply torch/testing/_asserts.py:206]: What do you think?

File: test/test_testing.py (hunk 1 above)
Comment (pmeier) [reply test/test_testing.py:864]: > I'm not sure I understand what you mean about changing the size?

My bad, I misread the diff.

File: torch/testing/_asserts.py [hunk 10]
```diff
@@ -154,58 +158,71 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
...
-    - abs (int): Absolute number of mismatches.
-    - rel (float): Relative number of mismatches.
-    - idx (Union[int, Tuple[int, ...]]): Index of greatest absolute difference.
-    - diff (Union[int, float]): Greatest absolute difference.
-    - a (Union[int, float]): Value of :attr:`a` at the greatest absolute difference.
-    - b (Union[int, float]): Value of :attr:`a` at the greatest absolute difference.
+    - total_elements (int): Total number of values.
+    - total_mismatches (int): Total number of mismatches.
+    - mismatch_ratio (float): Quotient of total mismatches and total elements.
+    - max_abs_diff (Union[int, float]): Greatest absolute difference of :attr:`a` and :attr:`b`.
```
Comment (mruberry) [inline torch/testing/_asserts.py:181]: This looks really good; we don't need to block this diff in the stack on this question, but is NaN vs 0 a greater absolute difference than inf vs 0? What about 5 vs 0?

File: torch/testing/_asserts.py [hunk 11]
```diff
@@ -154,58 +158,71 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
     return a, b
 
 
-_Trace = namedtuple("Trace", ("total", "abs", "rel", "idx", "diff", "a", "b"))
+_Trace = namedtuple(
```
Comment (mruberry) [inline torch/testing/_asserts.py:161]: This is cool, especially since it will be easy to extend in the future if we like

File: torch/testing/_asserts.py (hunk 8 above)
Comment (pmeier) [reply torch/testing/_asserts.py:206]: I would do it everywhere for consistency. Since the first two inputs should be positional only (although they can't be right now), this should not make it more verbose for the users. Plus, in case we want something asymmetric for `assert_equal` we are on the safe side.

File: torch/testing/_asserts.py [hunk 12]
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
...
+    Currently supports pairs of
+
+    - :class:`Tensor`'s,
+    - :class:`~collections.abc.Sequence`'s of :class:`Tensor`'s, and
+    - :class:`~collections.abc.Mapping`'s of :class:`Tensor`'s.
+
+    Args:
+        a (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): First input.
+        b (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Second input.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if two tensors match.
```
Comment (mruberry) [inline torch/testing/_asserts.py:387]: This is a nice design

File: torch/testing/_asserts.py [hunk 13]
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
...
+            In case they mismatch should return an :class:`Exception` with an expressive error message.
+
+    Returns:
+        (Optional[Exception]): :class:`UsageError` if the inputs types are unsupported. Additionally, any exception
+        returned by :attr:`check_tensors`.
+    """
+    # _check_are_tensors() returns nothing in case both inputs are tensors and an exception otherwise. Thus, the logic
+    # is inverted here.
+    are_tensors = not _check_are_tensors(a, b)
```
Comment (mruberry) [inline torch/testing/_asserts.py:396]: Might as well check explicitly for this being None

Is it weird that it doesn't return a tuple (True/False, optional exception)?

File: torch/testing/_asserts.py [hunk 14]
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
...
+        return AssertionError(f"The length of the sequences mismatch: {len_a} != {len_b}")
+    for idx, (a_t, b_t) in enumerate(zip(a, b)):
+        exc = check_tensors(a_t, b_t)
+        if exc:
+            return _amend_error_message(exc, f"{{}}\n\n{_SEQUENCE_MSG_FMTSTR.format(idx)}")
+
+    return None
+
+
+_MAPPING_MSG_FMTSTR = "The failure occurred for key '{}' of the mappings."
```
Comment (mruberry) [inline torch/testing/_asserts.py:459]: It's great to propagate the origins of the failures like this

File: torch/testing/_asserts.py [hunk 15]
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
...
+            first offending key.
+    """
+    a_keys = set(a.keys())
+    b_keys = set(b.keys())
+    if a_keys != b_keys:
+        missing_keys = b_keys - a_keys
+        additional_keys = a_keys - b_keys
+        return AssertionError(
+            f"The keys of the mappings do not match:\n\n"
+            f"Missing keys in the first mapping: {sorted(missing_keys)}\n"
```
Comment (mruberry) [inline torch/testing/_asserts.py:486]: This is cool and we shouldn't block this diff on this question, but this might be more natural with "expected" and "actual" vs. "first"

File: torch/testing/_asserts.py [hunk 16]
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
...
-        check_device (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` live in the same
-            :attr:`~torch.Tensor.device` memory. If this check is disabled **and** :attr:`a` and :attr:`b` do not live
-            in the same memory :attr:`~torch.Tensor.device`, they are moved CPU memory before their values are
-            compared.
-        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
-            :attr:`~torch.Tensor.dtype`. If this check is disabled **and** :attr:`a` and :attr:`b` do not have the same
-            :attr:`~torch.Tensor.dtype`, they are copied to the :class:`~torch.dtype` returned by
-            :func:`torch.promote_types` before their values are compared.
-        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same stride.
+        check_device (bool): If ``True`` (default), asserts that tensors live in the same :attr:`~torch.Tensor.device`
```
Comment (mruberry) [inline torch/testing/_asserts.py:587]: nit: "that both tensors are on the same device" would be more consistent with other PyTorch language

Comment (mruberry) [reply torch/testing/_asserts.py:587]: (because tensors, like vampires, aren't really alive)

**Review by mruberry — APPROVED** [review]
Awesome!

At this point I think the issues with this diff are so minor or debatable that we should revisit them in future diffs to maintain code velocity.

For example, "actual" and "expected" is a simple refinement to perform later, and I don't want slow down the landing process for it.

**Comment by mruberry** [comment]
Would you push a trivial update to this branch, @pmeier, to retrigger the lint builds?

File: torch/testing/_asserts.py (hunk 10 above)
Comment (pmeier) [reply torch/testing/_asserts.py:181]:
```python
>>> torch.sort(torch.tensor([5, float("inf"), float("nan")])).values
tensor([5., inf, nan])
```

This means in case any NaN or Inf value is present it will always be used as greatest diff. We should definitively revisit this in the future.

File: torch/testing/_asserts.py (hunk 15 above)
Comment (pmeier) [reply torch/testing/_asserts.py:486]: Resolves automatically with https://github.com/pytorch/pytorch/pull/55145#discussion_r613042365

File: torch/testing/_asserts.py (hunk 13 above)
Comment (pmeier) [reply torch/testing/_asserts.py:396]: Returning `ok:bool, exc: Optional[Exception]` has redundant information since the output would always be one of `True, Exception` or `False, None`. Thus, `ok == not(bool(exc))`. This discussion is moot anyway, because #55786 removes this check in order to allow other input types.
